// waves-api/raw/src/classes/AssetPair.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("../api/matcher/index");
var Asset_1 = require("./Asset");
function getMatcherPairOrder(assetOne, assetTwo) {
    assetOne = Asset_1.default.isAsset(assetOne) ? assetOne.id : assetOne;
    assetTwo = Asset_1.default.isAsset(assetTwo) ? assetTwo.id : assetTwo;
    return index_1.v1.getOrderbook(assetOne, assetTwo).then(function (orderbook) {
        return {
            amountAssetId: orderbook.pair.amountAsset,
            priceAssetId: orderbook.pair.priceAsset
        };
    });
}
var AssetPair = /** @class */ (function () {
    function AssetPair(amountAsset, priceAsset) {
        if (!Asset_1.default.isAsset(amountAsset) || !Asset_1.default.isAsset(priceAsset)) {
            throw new Error('Please use instances of Asset in the AssetPair constructor');
        }
        this.amountAsset = amountAsset;
        this.priceAsset = priceAsset;
        this.precisionDifference = priceAsset.precision - amountAsset.precision;
    }
    AssetPair.prototype.toJSON = function () {
        return {
            amountAsset: this.amountAsset.id,
            priceAsset: this.priceAsset.id
        };
    };
    AssetPair.prototype.toString = function () {
        return this.amountAsset + "/" + this.priceAsset;
    };
    return AssetPair;
}());
exports.default = {
    get: function (assetOne, assetTwo) {
        return getMatcherPairOrder(assetOne, assetTwo).then(function (matcherPair) {
            return Promise.all([
                Asset_1.default.get(matcherPair.amountAssetId),
                Asset_1.default.get(matcherPair.priceAssetId)
            ]);
        }).then(function (assets) {
            return new AssetPair(assets[0], assets[1]);
        });
    },
    isAssetPair: function (object) {
        return object instanceof AssetPair;
    }
};
