// waves-api/raw/src/classes/Money.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var bignumber_1 = require("../libs/bignumber");
var Asset_1 = require("./Asset");
// TODO : stop exporting the following functions when OrderPrice is removed
function checkAmount(amount) {
    if (!(typeof amount === 'string' || amount instanceof bignumber_1.default)) {
        throw new Error('Please use strings to create instances of Money');
    }
}
exports.checkAmount = checkAmount;
function getDivider(precision) {
    return new bignumber_1.default(10).pow(precision);
}
exports.getDivider = getDivider;
var Money = /** @class */ (function () {
    function Money(coins, asset) {
        this.asset = asset;
        this.coins = new bignumber_1.default(coins);
        this.divider = getDivider(asset.precision);
    }
    Money.prototype.toCoins = function () {
        return this.coins.toFixed(0);
    };
    Money.prototype.toTokens = function () {
        return this.coins.div(this.divider).toFixed(this.asset.precision);
    };
    Money.prototype.toJSON = function () {
        return {
            assetId: this.asset.id,
            tokens: this.toTokens()
        };
    };
    Money.prototype.toString = function () {
        return this.toTokens() + " " + this.asset.id;
    };
    return Money;
}());
exports.default = {
    fromCoins: function (coins, supposedAsset) {
        checkAmount(coins);
        return Asset_1.default.get(supposedAsset).then(function (asset) {
            return new Money(coins, asset);
        });
    },
    fromTokens: function (tokens, supposedAsset) {
        checkAmount(tokens);
        return Asset_1.default.get(supposedAsset).then(function (asset) {
            var divider = getDivider(asset.precision);
            tokens = new bignumber_1.default(tokens).toFixed(asset.precision);
            var coins = new bignumber_1.default(tokens).mul(divider);
            return new Money(coins, asset);
        });
    },
    isMoney: function (object) {
        return object instanceof Money;
    }
};
